File: Sharp/Scanner.cpp
#include "Scanner.hpp"

#include <cctype>
#include <exception>
#include <iostream>
#include <map>

#include "ScanException.hpp"

static const std::map<const char *, TokenType> Keywords = 
    {
        {"AND", TokenType::AND},
        {"CLASS", TokenType::CLASS},
        {"ELSE", TokenType::ELSE},
        {"FALSE", TokenType::FALSE},
        {"FUN", TokenType::FUNCTION},
        {"FOR", TokenType::FOR},
        {"IF", TokenType::IF},
        {"NULL", TokenType::NUL},
        {"OR", TokenType::OR},
        {"RETURN", TokenType::RETURN},
        {"SUPER", TokenType::SUPER},
        {"THIS", TokenType::THIS},
        {"TRUE", TokenType::TRUE},
        {"VAR", TokenType::VAR},
        {"WHILE", TokenType::WHILE},
        {"VOID", TokenType::VOID}
    };

Scanner::Scanner(std::istream& stream) :
    line(1),
    column(0),
    stream(stream),
    tokens(std::vector<Token>())
{
}

Scanner::~Scanner()
{
}

std::vector<Token> Scanner::GetTokens(std::istream& stream)
{
    Scanner scanner = Scanner(stream);
    scanner.Scan();
    return scanner.tokens;
}

bool Scanner::Match(char c)
{
    if (this->stream.eof()) 
        return false;
    if (this->stream.peek() != c) 
        return false;
    this->Advance();
    return true;
}

char Scanner::Advance()
{
    if (this->stream.eof()) 
        return 0;
    this->column++;
    return (char)this->stream.get();
}

void Scanner::AddToken(TokenType&& token)
{
    this->AddToken(std::move(token), std::move(Literal()));
}

void Scanner::AddToken(TokenType&& token, int lineStart, int columnStart)
{
    this->AddToken(std::move(token), std::move(Literal()), lineStart, columnStart);
}

void Scanner::AddToken(TokenType&& token, Literal literal)
{
    this->tokens.push_back(Token(std::move(token), 
        literal, line, column));
}

void Scanner::AddToken(TokenType&& token, Literal literal, int lineStart, int columnStart)
{
    this->tokens.push_back(Token(std::move(token), 
        literal, lineStart, columnStart));
}

void Scanner::PutBack(char c)
{
    this->stream.putback(c);
    this->column--;
}

void Scanner::Scan()
{
    while (!this->stream.eof()) {
        this->ScanToken();
    }
}

void Scanner::ScanIdentifier(char c)
{
    char currChar = c;
    int lineStart = this->line;
    int columnStart = this->column;
    std::stringstream ssOrig;
    std::stringstream ssUpper;
    while (isalnum(currChar) || currChar == '_')
    {
        ssOrig << currChar;
        ssUpper << toupper(currChar);
        currChar = Advance();
    }
    this->PutBack(currChar);
    auto it = Keywords.find(ssUpper.str().c_str());
    if (it == Keywords.end())
    {
        this->AddToken(TokenType::IDENTIFIER, 
            ssOrig.str().c_str(), lineStart, columnStart);
    }
    else 
    {
        TokenType type = it->second;
        this->AddToken(std::move(type), lineStart, columnStart);
    }
}

void Scanner::ScanNumber(char c)
{
    char currChar = c;
    int lineStart = this->line;
    int columnStart = this->column;
    std::stringstream ss;
    while (isdigit(currChar) || currChar == ',' || currChar == '.')
    {
        ss << currChar;
        currChar = Advance();
    }
    this->PutBack(currChar);
    this->AddToken(TokenType::NUMBER, 
        Literal(ss.str().c_str()), lineStart, columnStart);
}

void Scanner::ScanString()
{
    int lineStart = this->line;
    int columnStart = this->column;
    char currChar = this->Advance();
    std::stringstream ss;
    while (currChar != '"' && !this->stream.eof())
    {
        ss << currChar;
        currChar = this->Advance();
    }
    if (this->stream.eof())
    {
        throw ScanException("Unterminated string.", lineStart, columnStart);
    }
    this->AddToken(TokenType::STRING, 
        Literal(ss.str().c_str()), lineStart, columnStart);
}

void Scanner::ScanToken()
{
    char c = this->Advance();
    switch (c) {
    case ('('):
        this->AddToken(TokenType::LEFT_PAREN);
        break;
    case (')'):
        this->AddToken(TokenType::RIGHT_PAREN);
        break;
    case ('{'):
        this->AddToken(TokenType::LEFT_BRACE);
        break;
    case ('}'):
        this->AddToken(TokenType::RIGHT_BRACE);
        break;
    case ('.'):
        this->AddToken(TokenType::DOT);
        break;
    case (','):
        this->AddToken(TokenType::COMMA);
        break;
    case ('+'):
        this->AddToken(TokenType::PLUS);
        break;
    case ('-'):
        this->AddToken(TokenType::MINUS);
        break;
    case ('#'):
        this->AddToken(TokenType::HASH);
        break;
    case (':'):
        this->AddToken(TokenType::COLON);
        break;
    case (';'):
        this->AddToken(TokenType::SEMICOLON);
        break;
    case ('*'):
        this->AddToken(TokenType::STAR);
        break;
    case ('!'):
        this->AddToken(this->Match('=') ? 
            TokenType::BANG_EQUAL : TokenType::BANG);
        break;
    case ('='):
        this->AddToken(this->Match('=') ? 
            TokenType::EQUAL_EQUAL : TokenType::EQUAL);
        break;
    case ('<'):
        this->AddToken(this->Match('=') ? 
            TokenType::LESS_EQUAL : TokenType::LESS);
        break;
    case ('>'):
        this->AddToken(this->Match('=') ? 
            TokenType::GREATER_EQUAL : TokenType::GREATER);
        break;
    case ('/'):
        if (Match('/')) 
        {
            while ((this->stream.peek() != '\n' && !this->stream.eof())) 
            {
                this->Advance();
            }
        }
        else 
        {
            AddToken(TokenType::SLASH);
        }
        break;
    case (' '):
        break;
    case ('\r'):
    case ('\n'):
    case ('\t'):
        this->line++;
        this->column = 0;
        break;
    case ('"'):
        this->ScanString();
        break;
    default:
        if (isdigit(c))
        {
            this->ScanNumber(c);
        }
        else if (isalpha(c) || c == '_')
        {
            this->ScanIdentifier(c);
        }
        else 
        {
            if (this->stream.eof())
                return;
            std::stringstream ss;
            ss << "Unexpected character '" << c << "'" << std::endl;
            throw ScanException(ss.str(), this->line, this->column);
        }
    }
}

File: Sharp/Scanner.cs
namespace Sharp
{
    public class Scanner
    {
        private string _source;
        private List<Token> _tokens = new List<Token>();
        private int _start = 0;
        private int _current = 0;
        private int _line = 0;
        private Dictionary<string, TokenType> _keywords = new Dictionary<string, TokenType>() 
        {
            { "and",    TokenType.AND },
            { "class",  TokenType.CLASS },
            { "else",   TokenType.ELSE },
            { "false",  TokenType.FALSE },
            { "for",    TokenType.FOR },
            { "fun",    TokenType.FUN },
            { "if",     TokenType.IF },
            { "nil",    TokenType.NIL },
            { "or",     TokenType.OR },
            { "print",  TokenType.PRINT },
            { "return", TokenType.RETURN },
            { "super",  TokenType.SUPER },
            { "this",   TokenType.THIS },
            { "true",   TokenType.TRUE },
            { "var",    TokenType.VAR },
            { "while",  TokenType.WHILE },
        };

        public Scanner(string source)
        {
            _source = source;
        }

        public List<Token> ScanTokens()
        {
            while (!IsAtEnd())
            {
                _start = _current;
                ScanToken();
            }
            return _tokens;
        }

        private void AddToken(TokenType type, object? literal = null)
        {
            _tokens.Add(new Token(type, _source.Substring(_start, _current), literal, _line));
        }

        private void ScanToken() {
            char c = Advance();
            switch (c)
            {
                case '(': AddToken(TokenType.LEFT_PAREN); break;
                case ')': AddToken(TokenType.RIGHT_PAREN); break;
                case '{': AddToken(TokenType.LEFT_BRACE); break;
                case '}': AddToken(TokenType.RIGHT_BRACE); break;
                case ',': AddToken(TokenType.COMMA); break;
                case '.': AddToken(TokenType.DOT); break;
                case '-': AddToken(TokenType.MINUS); break;
                case '+': AddToken(TokenType.PLUS); break;
                case ';': AddToken(TokenType.SEMICOLON); break;
                case '*': AddToken(TokenType.STAR); break;
                case '!':
                    AddToken(Match('=') ? TokenType.BANG_EQUAL : TokenType.BANG);
                    break;
                case '=':
                    AddToken(Match('=') ? TokenType.EQUAL_EQUAL : TokenType.EQUAL);
                    break;
                case '<':
                    AddToken(Match('=') ? TokenType.LESS_EQUAL : TokenType.LESS);
                    break;
                case '>':
                    AddToken(Match('=') ? TokenType.GREATER_EQUAL : TokenType.GREATER);
                    break;
                case '/':
                    // If this is a comment
                    if (Match('/'))
                    {
                        // skip until the end of line
                        while (Peek() != '\n' && !IsAtEnd())
                        {
                            Advance();
                        }
                    }
                    else
                    {
                        AddToken(TokenType.SLASH);
                    }
                    break;
                case ' ':
                case '\r':
                case '\t':
                    break;
                case '\n':
                    _line++;
                    break;
                case '"':
                    GetString();
                    break;
                default:
                    if (IsDigit(c))
                    {
                        GetNumber();
                    }
                    else if (IsAlpha(c))
                    {
                        GetIdentifier();
                    }
                    else
                    {
                        Sharp.Error(_line, "Unexpected character '" + c + "'.");
                    }
                    break;
            }
        }

        private bool IsAtEnd()
        {
            return _current >= _source.Length;
        }

        private bool IsDigit(char c)
        {
            return c >= '0' && c <= '9';
        }

        private bool IsAlpha(char c)
        {
            return ((c >= 'a' && c <= 'z') ||
                    (c >= 'A' && c <= 'Z') ||
                    (c == '_'));
        }

        private bool IsAlphaNumeric(char c)
        {
            return IsDigit(c) || IsAlpha(c);
        }

        private char Advance()
        {
            return _source.ElementAt(_current++);
        }

        private char Peek()
        {
            if (IsAtEnd())
                return '\0';
            return _source.ElementAt(_current);
        }

        private char PeekNext()
        {
            if (_current + 1 >= _source.Length)
                return '\0';
            return _source.ElementAt(_current + 1);
        }

        private void GetString()
        {
            while (Peek() != '"' && !IsAtEnd())
            {
                if (Peek() == '\n')
                {
                    _line++;
                }
                Advance();
            }
            if (IsAtEnd())
            {
                Sharp.Error(_line, "Unterminated string.");
            }
            Advance();
            AddToken(TokenType.STRING, _source.Substring(_start + 1, _current - 1));
        }

        private void GetNumber()
        {
            while (IsDigit(Peek()))
            {
                Advance();
            }
            if (Peek() == '.' && IsDigit(PeekNext()))
            {
                Advance();
            }
            while (IsDigit(Peek()))
            {
                Advance();
            }
            AddToken(TokenType.NUMBER, Double.Parse(_source.Substring(_start, _current)));
        }

        private void GetIdentifier()
        {
            while (IsAlphaNumeric(Peek()))
            {
                Advance();
            }
            var text = _source.Substring(_start, _current);
            if (_keywords.ContainsKey(text))
            {
                AddToken(_keywords[text]);
            }
            else
            {
                AddToken(TokenType.IDENTIFIER);
            }
        }

        private bool Match(char expectedChar)
        {
            if (IsAtEnd())
            {
                return false;
            }
            if (_source.ElementAt(_current) != expectedChar)
            {
                return false;
            }
            _current++;
            return true;
        }
    }
}